streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingTimestampedWindowBytesStore.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.state.WindowStore;
+
+import static org.apache.kafka.streams.state.internals.ValueAndTimestampDeserializer.rawValue;
+import static org.apache.kafka.streams.state.internals.ValueAndTimestampDeserializer.timestamp;
+
+class ChangeLoggingTimestampedWindowBytesStore extends ChangeLoggingWindowBytesStore {
+
+    ChangeLoggingTimestampedWindowBytesStore(final WindowStore<Bytes, byte[]> bytesStore,
+                                             final boolean retainDuplicates) {
+        super(bytesStore, retainDuplicates);
+    }
+
+    @Override
+    void log(final Bytes key,
+             final byte[] valueAndTimestamp) {
+        if (valueAndTimestamp != null) {
+            changeLogger.logChange(key, rawValue(valueAndTimestamp), timestamp(valueAndTimestamp));
+        } else {
+            changeLogger.logChange(key, null);
```

Comment:
This is not specifically related to this line of code but when I read about this, I was thinking about some correlated topic: when we call `kvStore.delete` and `sessionStore.remove`, we will pass `null` as ValueAndTimestamp and hence the timestamp field will also be null. Whereas in `windowStore` we do not have a delete api, but users can call `put(k, null)` or `put(k, null, timestamp)` in which case the timestamp field will not be null actually. Hence there's a discrepancy between the first two and the latter one, right?

Reply:
`WindowStore#put(k, null, timestamp)` is more or less the same as `put(k, null)` (the later actually calls the former and used `context.timestamp()` as third parameter) -- note that the timestamp that is passed is the `windowStartTimestamp` (not the timestamp that is associated with the value for the timestamped store).

In fact, we plan to deprecate and remove `put(k, null)`, because it does not make sense semantically, to put something into a window store without specifying the window start timestamp (and `context.timestamp()` is not a good default value). Thus, for `put(k, null, timestamp)` the provided timestamp would be ignored anyway.

Also note, that for timestamped-store, the call would be `put(key, valueAndTimestamp, windowStartTimestamp)` and the timestamp in the value is not related to the window-start-timestamp (that will be part of the key in the store).

I agree that calling `put(key, null, timestamp)` to delete a window would be weird. However, in the DSL we would never delete a window anyway and don't need this. Not sure if PAPI uses would want a proper `remove(key)` method.

Reply:
We have many gaps in store documentation, and I agree it would be worth to do a follow up for this.

I also think I understand now why `WindowStore#put(k, v)` was added in the first place. For stream-steam join, we use windowed stores (with "allow duplicates" enabled) to store each individual record and thus there is not really a notion of a window (we use it more like a key-value store that allows for time-range queried plus retention time), and we use the record timestamp as "window start timestamp" (ie, we can use this shortcut method). Still think, we should remove it though.

Reply:
Thanks for the explanation @mjsax , yes we were trying to "reuse" the window store for stream-stream windowed join, which, as an after-thought is not a very good design.

I think we do not need to make any code changes atm (as for doc changes maybe we can track those as a separate ticket / PR).